crates/stackable-certs/src/cert_builder.rs

```diff
@@ -0,0 +1,330 @@
+    keys::CertificateKeypair,
+};
+
+/// Defines all error variants which can occur when creating a certificate
+#[derive(Debug, Snafu)]
+pub enum CreateCertificateError<E>
+where
+    E: std::error::Error + 'static,
+{
+    #[snafu(display("failed to parse validity"))]
+    ParseValidity { source: x509_cert::der::Error },
+
+    #[snafu(display("failed to parse \"{subject}\" as subject"))]
+    ParseSubject {
+        source: x509_cert::der::Error,
+        subject: String,
+    },
+
+    #[snafu(display("failed to create key pair"))]
+    CreateKeyPair { source: E },
+
+    #[snafu(display("failed to serialize public key as PEM"))]
+    SerializePublicKey { source: x509_cert::spki::Error },
+
+    #[snafu(display("failed to decode SPKI from PEM"))]
+    DecodeSpkiFromPem { source: x509_cert::der::Error },
+
+    #[snafu(display("failed to create certificate builder"))]
+    CreateCertificateBuilder { source: x509_cert::builder::Error },
+
+    #[snafu(display("failed to add certificate extension"))]
+    AddCertificateExtension { source: x509_cert::builder::Error },
+
+    #[snafu(display("The subject alternative DNS name \"{dns_name}\" is not a Ia5String"))]
+    SaDnsNameNotAIa5String {
+        dns_name: String,
+        source: x509_cert::der::Error,
+    },
+
+    #[snafu(display("failed to build certificate"))]
+    BuildCertificate { source: x509_cert::builder::Error },
+}
+
+/// This builder builds certificates of type [`CertificatePair`].
+///
+/// Example code to construct a certificate:
+///
+/// ```no_run
+/// use stackable_certs::{
+///     keys::ecdsa,
+///     ca::{CertificateAuthority, CertificateAuthorityBuilder},
+///     CertificateBuilder,
+/// };
+///
+/// let ca: CertificateAuthority<ecdsa::SigningKey> = CertificateAuthorityBuilder::builder()
+///     .build()
+///     .build_ca()
+///     .expect("failed to build CA");
+///
+/// let certificate = CertificateBuilder::builder()
+///     .subject("CN=trino-coordinator-default-0")
+///     .signed_by(&ca)
+///     .build()
+///     .build_certificate()
+///     .expect("failed to build certificate");
+/// ```
+#[derive(Builder)]
+pub struct CertificateBuilder<'a, KP>
+where
+    KP: CertificateKeypair,
+    <KP::SigningKey as signature::Keypair>::VerifyingKey: EncodePublicKey,
+{
+    /// Required subject of the certificate, usually starts with `CN=`.
+    subject: &'a str,
+
+    /// Optional list of subject alternative names (SAN) DNS entries,
+    /// that are added to the certificate.
+    #[builder(default)]
+    subject_alterative_dns_names: &'a [&'a str],
+
+    /// Validity/lifetime of the certificate.
+    ///
+    /// If not specified the default of [`DEFAULT_CERTIFICATE_VALIDITY`] will be used.
+    #[builder(default = DEFAULT_CERTIFICATE_VALIDITY)]
+    validity: Duration,
+
+    /// Serial number of the generated certificate.
+    ///
+    /// If not specified a random serial will be generated.
+    serial_number: Option<u64>,
```

Comment:
note: Again, I would like this to be an internal details not exposed to the user.

Reply:
Same as https://github.com/stackabletech/operator-rs/pull/1044#discussion_r2108327741